distribution/lib/Standard/Database/0.0.0-dev/src/Data/Internal/Base_Generator.enso

```diff
@@ -190,6 +190,8 @@ generate_expression dialect expr = case expr of
         op = dialect.operation_map.get_or_else kind (Error.throw <| Unsupported_Database_Operation_Error kind)
         parsed_args = arguments.map (here.generate_expression dialect)
         op parsed_args
+    IR.Order_Descriptor col_expr direction _ _ ->
+        Pair (here.generate_expression dialect col_expr) direction
```

Comment:
I propose the following solution: we can modify `IR.Operation` to contain two kinds of arguments 'expression_arguments' and other 'parameters'. Expression arguments would be passed through the `generate_expression`, while the other parameters would be passed to each function generator as-is. Then we can feed the order expressions as pure expressions and the sort direction parameters can be fed as this separate set of arguments.

This would involve a relatively large refactor as we now would need to make all operation generators take this second set of arguments (even though in most cases it will be empty).  But, personally, I think it is worth doing to ensure a clean architecture. Unless you expect a big rewrite to these parts soon, but I don't think we plan this?

This would actually help in a few other places too - for example we could remove the `FIRST` and `FIRST_NOT_NULL` variants and just make ignore null an 'other parameter', and the same for the percentiles in the PERCENTILE operation which are currently encoded as an `IR.Constant` but they could just be a regular constant.

Reply:
An alternative, possibly even cleaner solution could be to replace the `IR.Operation` atom with an ADT where we would have a constructor for _every_ SQL operation we support. For some reason initially I avoided this approach, mostly because I thought it's not very extensible and I was thinking of having some "automatic" operations, i.e. generating some default SQL code in a given dialect without even containing a mapping for that operation in that specific dialect. But this never happened. In retrospect, I think this would be the cleanest solution as it would allow as to control the arity of each operation (and allow to remove the prevalent `if args.length < 2 then Error.throw (Illegal_State_Error "Insufficient number of arguments for the operation.") else` cases). This is a bigger refactor, although it is quite mechanical.

If we decide it's worth it, I think I should be able to perform it in around one day.